### ISSUE TITLE
self hosted ASP.NET Core Web Api capable of starting and stopping causing memory leaks

### ISSUE THREAD

**Issue by kiiiieeeeuuuuw** [CoreWCF/CoreWCF#940]
I'm attempting to convert an old WCF service to an ASP.NET Core Web API, making use of the CoreWCF package. A key feature of this existing service is that it's being self hosted by an other application and is able to gracefully start & stop, without creating memory leaks. 

I have been able to figure out how to start and stop a prototype service. However, after performing some stress testing, it does seem like I've left a memory leak somewhere and I'm sadly out of ideas or available documentation at this point. I'm also considering that an ASP.NET Core Web API just isn't supposed to be used like this and I misunderstood this, if so, be sure to let me know. Also my apologies for the truckload of code, but I'm not sure what's relevant or not to the question.

The code for my prototype service looks like this:

Configuring the webhost:

``` 
private void CreateWebHostBuilder(){
    host = new WebHostBuilder() //WebHost.CreateDefaultBuilder()
                .UseKestrel(options =>
                {
                    options.AllowSynchronousIO = true;
                    options.ListenLocalhost(Startup.PORT_NR);
                    options.ListenLocalhost(Startup.HTTPS_PORT_NR, listenOptions => {
                        listenOptions.UseHttps("{self_signed certificate}", "{password}");
                    });
                })
                .UseShutdownTimeout(TimeSpan.FromSeconds(1))
                .UseStartup<Startup>()
                .Build();
                var mb = host.Services.GetRequiredService<ServiceMetadataBehavior>();
                mb.HttpsGetEnabled = true;
```
Inside the ```Startup``` class:

Configuring the ```IApplicationBuilder```:
```
public void Configure(IApplicationBuilder app){
    app.UseServiceModel(builder =>
    {
        // Add the Echo Service
        builder.AddService<EchoService>()                
        // Add service web endpoint                
        .AddServiceWebEndpoint<EchoService, IEchoService>(
            "echo",behavior => { behavior.HelpEnabled = true;}
        );
     });            
     app.UseMiddleware<SwaggerMiddleware>();            
     app.UseSwaggerUI();
     app.UseAuthentication();
}
```

Configuring the services:

```
public void ConfigureServices(IServiceCollection services){
    services.AddServiceModelServices();
    services.AddServiceModelWebServices();
    services.AddHostedService<EchoService>();
    services.AddSingleton(new SwaggerOptions());
    services.AddAuthentication(CertificateAuthenticationDefaults.AuthenticationScheme);          
}
```

The service interface:
```
[ServiceContract]
[OpenApiBasePath($"/{Startup.WEB_API_PATH}")]
public interface IEchoService : IHostedService {
    [OperationContract]
    [WebGet(UriTemplate = "/hello")]
    [OpenApiOperation(Description = "Method used to receive a friendly \"Hello world\"",
        Summary = "Hello world")]
    [OpenApiResponse(Description = "OK Response", StatusCode = HttpStatusCode.OK)]
    string HelloWorld();        
}
```

The implemented service:
```
public class EchoService : IEchoService {        
    public EchoService() { }

    public string HelloWorld() {
        return "Hello world!";
    }        
    public Task StartAsync(CancellationToken cancellationToken) {
        return Task.CompletedTask;
    }

    public Task StopAsync(CancellationToken cancellationToken)
    {            
        return Task.CompletedTask;
    }        
}
```

Creating and starting the host + services:
```
public void StartWebService(object obj){
    CreateWebHostBuilder();
    host.StartAsync();
}
```

Stopping and disposing the services and host:
```
public void StopWebService(object obj) {
    host.StopAsync().Wait();
    host.Dispose();                            
}
```

So if anyone has any suggestions or tutorial reference, be sure to let me know, any help is welcome.

**Comment by mconnew** [comment]
Starting and stopping it repeatedly should work without a problem. Would you be able to create a complete small test project which demonstrates the memory leak and I should be able to track it down relatively quickly.  
An additional validation step is to run `GC.Collect(2);` twice between each start/stop iteration. In the past I've seen situations where simply having plenty of RAM on the machine and so the GC wasn't very aggressive at cleaning up was mistaken for a memory leak. Doing two collections (2 are needed because of finalization) would eliminate that possibility.

**Comment by mconnew** [comment]
The suggestion of GC.Collect wasn't intended for use in production code but as a diagnostic step. If you have plenty of memory, there's no hurry to clean things up. A UI app will be using the workstation GC which has different characteristics than the server GC that an app using the Web SDK will use which will make a difference. I'll take a look later and see if there's anything that stands out straight away.

**Comment by mconnew** [comment]
There are a couple of problems in your app. In App.xaml you had the wrong .NET namespace for `xmlns:local`. Also you are calling `host.StartAsync()` without waiting for the start to complete. This might be causing you additional problems.  
I've narrowed the memory leak to issues in asp.net core. Surprisingly the effect isn't as bad in a WPF app as it is in a console app. There something which looks like it's a memory leak in CoreWCF but it actually isn't. We coalesce CancellationTokenSource instances to prevent thrashing of the TimerManager. This means we can't track the end lifetime of them so they stay around until their cancellation timer fires. But we coalesce them so we don't have multiple firing at once (which is the point of coalescing them). I've drilled into some of the memory leaks and I've found 3 categories.  

There's two SocketException's thrown for each shutdown as there's actually two listening sockets, one for IPv4 and one for IPv6. These seem to be pinned with a strong handle, and there's some other structures like overlappeds related to it which are all still pinned.  

There's some file system watchers registered, I believe for handling configuration files changing at runtime. These aren't unregistered when the service stops.  

Logging providers have registered callbacks with various components, eg a change tracking callback has been registered with LoggingEventSource to be notified when there's been a change in what events are being logged. This isn't unregistered when stopping the WebHost.  

I'll be opening an issue in the asp.net core repo with an even more simplified app and will link it here.

**Comment by kiiiieeeeuuuuw** [comment]
Well first of all, thanks a lot for taking such deep dive into all of this and explaining everything so clearly.

The wrong .NET namespace for `xmlns:local` is a copy-paste mistake from making the project trying to isolate the memory leak from the base project. So this shouldn't be the cause for any of the initial issues.

I'll be sure to follow up on the issue in the asp.net core repo. I'm simply amazed by how much you've found on such a short amount of time. Do you have any tutorials on whatever you did to check all of this out, because memory leaks have definitly been the bane of my developer-existence for quite some time?

**Comment by mconnew** [comment]
The issue I opened is dotnet/aspnetcore#45579  
I worked for many years exclusively on performance of WCF in .NET Framework (about 6 years) before starting to work on the .NET Core WCF code base, and that was about 6 years ago. It's a bit difficult to impart 12 years of experience in a comment, but I can give a few pointers. I used 2 different tools. First I started by using VS diagnostic tools two take two heap snapshots. Here's the gui and where the button is:
![image](https://user-images.githubusercontent.com/8648390/207454548-37890468-c49b-49fc-824f-fd1f445f2ff7.png)
You click the button to get a snapshot, run the scenario which is leaking memory and click the button to take a second snapshot. Open the second snapshot and in the top right corner you can choose the baseline to compare with. Pick the first snapshot as the baseline and it will tell you the difference in counts between the two snapshots. It will also tell you the roots of the objects. This can sometimes be a little tricky to work out, especially with things like asp.net core. This is where the second tool comes in. So first you need to take a dump of the process. It might be worth taking the second snapshot while you have broken into the process so that the snapshot and the dump are reflecting the same thing. The second tool [I use is perfview](https://github.com/microsoft/perfview/releases). Run this tool and from the Memory menu choose "Take Heap Snapshot From Dump". You'll need to paste in the path to the dump file and enter a path for the .gcdump file you'll generate. Once generated, open the file and go to the Heap Stacks view. I usually clear the GroupPats field, but that's largely because I do a lot of work with framework code and the GroupPats are designed to hide that and only show you your own stuff. Your mileage may vary, might be worth trying it both ways. Anyway, this view gives you a tree view of the path to root on a statistical basis on a per object basis. Sometimes there can be multiple paths, in which case perfview attempts to make an educated guess to which is the path which is keeping the object alive in the heap and will bold that path. This is the key difference between using perfview and using VS to view the heap. This made is a lot easier to see what was rooting the objects. In VS I kept hitting cycles in the object graph and had a hard time finding the actual root.  

PerfView is an incredibly useful tool to add to your collection. For example, if you use it to collect events, you can turn off stack sampling (used for profiling code and is expensive both in CPU and disk space) and it will still collect lots of useful things like every exception call stack that happens. It can capture allocation call stacks so you can find out which methods are allocating objects, and give you counts on the stacks. The help built in is really good, but it takes quite a bit of using to really get a hang of visualizing things. But once you do, so many problems become really easy to diagnose with perfview.

**Comment by kiiiieeeeuuuuw** [comment]
Hi Matt, thanks for the pointers, I'll be sure to take these to heart. Even though I'lll aggree that it's impossible to impart 12 years of experience, I'm sure these tips will at least give me a 2 year shortcut. Thanks for all the work you've already done, I'll definitly follow the conversation on the asp.net repo and check out perfview to see if I can find the same things you pointed out.
